### ISSUE TITLE
Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded

### ISSUE THREAD

**Issue by jimcdonald** [LiuLabUB/HMMRATAC#10]
One of my samples won't complete all the jobs within the time allotted while the other jobs do finish. 

**Here is the error:**
[jimcdonald@login4 ATACseq.HMMR.STDERR.STDOUT.files]$ cat HMMR.v1.2.1.blacklist.A2780.ATACseq.library001.treat1.mock.3535969.err
Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded
        at WigMath.pileup.toBedGraph(pileup.java:192)
        at WigMath.pileup.build(pileup.java:184)
        at WigMath.pileup.<init>(pileup.java:50)
        at HMMR_ATAC.Main_HMMR_Driver.main(Main_HMMR_Driver.java:223)
slurmstepd: error: *** JOB 3535969 ON node079 CANCELLED AT 2019-01-08T11:46:06 DUE TO TIME LIMIT ***

**However**, I generally only get the slurmstepd message without the java exception. 

The SBATCH script I used to submit the samples is attached.
[HMMRATAC_GC_Troubleshooting_SBATCH.txt](https://github.com/LiuLabUB/HMMRATAC/files/2789201/HMMRATAC_GC_Troubleshooting_SBATCH.txt)

I think this could just be a problem with the size of the sequencing library for this sample. Its BAM file is 12 GB. I have had successful runs with 8.4 GB and 7.2 GB BAM files, but the majority of my other samples are around 3-4 GB.

Is this the issue or is it some form of user error? Any ideas on how to fix it? Thank you for all your help!

**Comment by jimcdonald** [comment]
Well, I was thinking of creating a second issue for this sample, but I realized they were so close if not the same thing, I think I should include it here:

One of my other samples is also giving heap space errors. An example of the errors follows:
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:2219)
        at java.util.ArrayList.grow(ArrayList.java:242)
        at java.util.ArrayList.ensureExplicitCapacity(ArrayList.java:216)
        at java.util.ArrayList.ensureCapacityInternal(ArrayList.java:208)
        at java.util.ArrayList.add(ArrayList.java:440)
        at ATACFragments.FragPileupGen.buildTracks(FragPileupGen.java:318)
        at ATACFragments.FragPileupGen.<init>(FragPileupGen.java:71)
        at HMMR_ATAC.Main_HMMR_Driver.main(Main_HMMR_Driver.java:421)
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:2219)
        at java.util.ArrayList.grow(ArrayList.java:242)
        at java.util.ArrayList.ensureExplicitCapacity(ArrayList.java:216)
        at java.util.ArrayList.ensureCapacityInternal(ArrayList.java:208)
        at java.util.ArrayList.add(ArrayList.java:440)
        at ATACFragments.FragPileupGen.buildTracks(FragPileupGen.java:318)
        at ATACFragments.FragPileupGen.<init>(FragPileupGen.java:71)
        at HMMR_ATAC.Main_HMMR_Driver.main(Main_HMMR_Driver.java:421)
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.HashMap.createEntry(HashMap.java:901)
        at java.util.HashMap.addEntry(HashMap.java:888)
        at java.util.HashMap.put(HashMap.java:509)
        at ATACFragments.FragPileupGen.buildTracks(FragPileupGen.java:234)
        at ATACFragments.FragPileupGen.<init>(FragPileupGen.java:71)
        at HMMR_ATAC.Main_HMMR_Driver.main(Main_HMMR_Driver.java:421)
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:2219)
        at java.util.ArrayList.grow(ArrayList.java:242)
        at java.util.ArrayList.ensureExplicitCapacity(ArrayList.java:216)
        at java.util.ArrayList.ensureCapacityInternal(ArrayList.java:208)
        at java.util.ArrayList.add(ArrayList.java:440)
        at ATACFragments.FragPileupGen.buildTracks(FragPileupGen.java:318)
        at ATACFragments.FragPileupGen.<init>(FragPileupGen.java:71)
        at HMMR_ATAC.Main_HMMR_Driver.main(Main_HMMR_Driver.java:421)
Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded
        at java.lang.Integer.valueOf(Integer.java:642)
        at ATACFragments.FragPileupGen.buildTracks(FragPileupGen.java:234)
        at ATACFragments.FragPileupGen.<init>(FragPileupGen.java:71)
        at HMMR_ATAC.Main_HMMR_Driver.main(Main_HMMR_Driver.java:421)

Unlike above, the BAM file for this sample is only 7.2 GB and another sample with a 7.2 GB BAM file works just fine.

Is this some form of the memory leak that left open too many files showing up again since I am using a cluster? Am I attempting to run too many processes from one SBATCH script (see the attachment in my earlier post)? Why does it seem to only affect two of my samples and not the others?

Thank you again for creating the tool and helping troubleshoot some of these issues.

**Comment by EvanTarbell** [comment]
Can you try lowering the --window option?  The first case the error is being thrown by the codes that create the pileup and the second one is thrown during the generation of the decomposed signals.  In both cases, HMMRATAC splits the genome based on this parameter and calculates the signals across these windows.

**Comment by jimcdonald** [comment]
Thanks! I realized after writing the message that I had missed your troubleshooting tip in the Guide file. I tried using 1/20th the default setting as recommended in the --window <int> part of Section 2. It worked.

Also, in an attempt to be helpful, Section 2 under the --window option recommends 1/20th the default while Section 4 recommends 1/10th the default. It may help avoid confusion in the future to reconcile those recommendations.
